[wip] Update to use Twilio voice sdk 2.0

sarahcstringer opened the pull request.

Commit:
Activate virtualenv before running server.

Commit:
Add more explanation for example .env file.

Commit:
Add node_modules to gitignore.

Commit:
Fix token decoding error.

Commit:
Remove CDN and use npm to install Twilio voice sdk.

Commit:
Remove reference to deprecated capability tokens.

Commit:
Remove Device options that are now defaults in voice-sdk 2.0.

enableRingingState and fakeLocalDtmf default to True in
the 2.x versions of Twilio Voice SDK.

Commit:
Stop listening for device.on(ready) and instead listen for registered.

The Device object is now lazy about opening the signaling connection.
`new Device(...)` is synchronous and will not cause a signaling channel
to be opened, and the "ready" event won't be fired.

Commit:
Move connect and disconnect events to Call object instead of Device object.

These events are now fired for Call objects instead of Device objects.

Commit:
Rename connect objects to call objects.

No functional change -- in the Twilio voice sdk 2.0, internal
Connection objects were renamed to Call objects, so this is a
semantic change to be in line with the internals.

Commit:
Pass device to updateDevices functions.

These functions were previously not working because they did not have
access to the device.

Commit:
Change param key for outbound call payload.

This is to disambiguate the payload that Twilio sends for inbound
calls. The payload typically includes a "To" field already.

Commit:
Allow for incoming calls, and change payload from to to phone.

Commit:
Add npm install step to install command.

Commit:
Update README with new install steps.

Commit:
Handle incoming calls in the UI.

Commit:
Simplify if statement for outbound calls.

Commit:
Fix tests.

onlywade commented on app.py:
```diff
@@ -46,28 +47,37 @@ def token():
     resp = VoiceResponse()
-    if "To" in request.form and request.form["To"] != '':
-        dial = Dial(caller_id=os.environ['TWILIO_CALLER_ID'])
+    if request.form.get("To") and request.form["To"] == twilio_number:
```
I presume that having two distinct conditional checks here may be intentional, for purposes of clarity. Personally, though, I would consider consolidating them into a single conditional, to be more concise:

```
if request.form.get("To") == twilio_number:
```

Just a minor preference - either way reads okay.

sarahcstringer replied:
Good call, the two checks were redundant 👍 I'll update.

onlywade reviewed (commented):
Excellent work! Aside from the SDK upgrade, I dig the small improvements around comments, naming, and overall organization.

While I don't have much insight into the SDK, or Javascript best practices or style guidelines, etc, I can at least say that everything passes the sniff test as far as I'm concerned. Also, fwiw, the demo runs great using this branch and following the readme. 👍

Commit:
Remove redundant check in if condition.

Commit:
Show accept/reject incoming call buttons on incoming call.

Commit:
Add screenshots and additional instructions to README.

stern-shawn commented on static/quickstart.js:
```diff
@@ -4,8 +4,9 @@
   var outputVolumeBar = document.getElementById('output-volume');
   var inputVolumeBar = document.getElementById('input-volume');
   var volumeIndicators = document.getElementById('volume-indicators');
+  var device, outgoingCall;
```
`outgoingCall` doesn't seem to be used unless that's being hidden in the diff that I'm looking at, and is redeclared on line 60. Maybe it isn't needed here?
```suggestion
  const volumeIndicators = document.getElementById('volume-indicators');<br class="Apple-interchange-newline">
  let device;
```

Side note: maybe it's outside the scope of this change, but I don't think we should demonstrate using `var` in our JS code samples since it's function scoped and generally unwieldy. Most of these variables can be declared using `const` instead, and a small number like `device` can be declared with `let` since they're mutable 😄 

Happy to bash on this a bit if that'd help~

stern-shawn commented on static/quickstart.js:
```diff
@@ -77,30 +45,20 @@
-    if (device) {
-      device.disconnectAll();
+      var outgoingCall = device.connect({params: params});
+      outgoingCall.then(callEstablished);
```
Two teeny nits:

- We can use object shorthand here to reduce the noise of `params = params`
- If `outgoingCall` isn't being referenced again later, and `connect` returns a promise, might be simpler to just chain immediately and not store `outgoingCall`
```suggestion
      device.connect({ params }).then(callEstablished);
```

stern-shawn commented on static/quickstart.js:
```diff
@@ -38,32 +30,8 @@
-          conn.accept();
-        }
+      device.on("incoming", function (call) {
+          incomingCallUI(call);
```
Similar shortcut to the trick you are doing on line 61 😃 
```suggestion
      device.on("incoming", incomingCallUI);
```

stern-shawn reviewed (commented):
Love this and the detailed readme changes!

I can't speak to the usage of the Twilio client itself, but I have some small JS-specific nits to toss into the ring 😆

Commit:
Use let/const instead of var.

Commit:
Simplify js code.

stern-shawn commented on static/quickstart.js:
```diff
@@ -110,7 +65,7 @@
 
   speakerDevices.addEventListener("change", function () {
-    var selectedDevices = [].slice
+    const selectedDevices = [].slice
```
Looks like this got dredged up by the var -> const change, haha.

`[].slice.call` is a pretty esoteric way to convert a `NodeList` into an `Array`. Would recommend using the more readable/beginner-friendly `Array.from` method. [MDN link](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from#array_from_a_nodelist)

```suggestion
    const selectedDevices = Array.from(speakerDevices.children)
      .filter((node) => node.selected); // the filter function can also be shortened :)
```

stern-shawn commented on static/quickstart.js:
```diff
@@ -123,7 +78,7 @@
 
   ringtoneDevices.addEventListener("change", function () {
-    var selectedDevices = [].slice
+    const selectedDevices = [].slice
```
Same as the other comment on `[].slice.call`

Commit:
Fix incorrect var initialization.

Commit:
Use more canonical JS for creating new arrays.

Commit:
Clarify comments in .env.example.

Remove example of creating a new API key via the twilio cli,
because doing so requires retrieving the account's auth token.

Also, make a note that the API_KEY is the key's SID.

Commit:
Fix assignment to a constant error.

Commit:
Lint python code.

Commit:
Pull in changes to frontend code from node quickstart.

Brianna made structure changes to the frontend code to make
it more readable/understandable.
https://github.com/TwilioDevEd/voice-javascript-sdk-quickstart-node

This also lints the frontend code with prettier.js.

Commit:
Remove colon from index.html check.
